src/shared/auto-update-cache.ts

```diff
@@ -0,0 +1,58 @@
+import { TimeString } from '@types';
+import ms, { StringValue } from 'ms';
+type Timestamp = number;
+type CacheConstructorParams<Value> = {
+  calculate: () => Promise<Value>;
+  config: AutoUpdateCacheConfig;
+};
+
+const DEFAULT_RETRY_TIME: TimeString = '5M';
+
+export type AutoUpdateCacheConfig = {
+  valid?: 'always' | { onlyFor: StringValue };
+  update: { every: StringValue; ifFailsTryAgainIn?: StringValue };
+};
+
+export class AutoUpdateCache<Value> {
+  private readonly calculate: () => Promise<Value>;
+  private readonly config: AutoUpdateCacheConfig;
+  private cache: { lastUpdated: Timestamp; value: Value | undefined };
+
+  constructor({ calculate, config }: CacheConstructorParams<Value>) {
+    this.calculate = calculate;
+    this.config = config;
+    this.cache = { lastUpdated: 0, value: undefined };
+
+    const isInvalid = config.valid && config.valid !== 'always' && !config.valid.onlyFor;
```

Comment:
Not sure exactly what we are verifying here. Doesn't typescript make sure that the string is a "TimeString"?
Also, maybe we should make sure that if `onlyFor` is set, then it's greater than `update.every`